MSAL/test/integration/native_auth/end_to_end/otp_code_retriever/MSALNativeAuthEmailCodeRetriever.swift

```diff
@@ -0,0 +1,102 @@
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+import XCTest
+
+class MSALNativeAuthEmailCodeRetriever: XCTestCase {
+
+    private let baseURLString = "https://www.1secmail.com/api/v1/?action="
+    private let secondsToWait = 4.0
+    private let numberOfRetry = 3
+
+    func retrieveEmailOTPCode(email: String) async -> String? {
+        let comps = email.components(separatedBy: "@")
+        guard comps.count == 2 else {
+            XCTFail("Invalid email address")
+            return nil
+        }
+        let local = comps[0]
+        let domain = comps[1]
+
+        guard let lastMessageId = await retrieveLastMessage(local: local, domain: domain, retryCounter: numberOfRetry) else {
+            XCTFail("Something went wrong retrieving the messages for the email specified")
+            return nil
+        }
+        guard let emailOTPCode = await retrieveOTPCodeFromMessage(local: local, domain: domain, messageId: lastMessageId) else {
+            XCTFail("Something went wrong retrieving the OTP code")
+            return nil
+        }
+        return emailOTPCode
+    }
+
+    private func retrieveLastMessage(local: String, domain: String, retryCounter: Int) async -> Int? {
+        guard retryCounter > 0, let url = URL(string: baseURLString + "getMessages&login=\(local)&domain=\(domain)") else {
+            return nil
+        }
+        try? await Task.sleep(nanoseconds: UInt64(secondsToWait * Double(NSEC_PER_SEC)))
+        var request = URLRequest(url: url)
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        do {
+            let (data, response) = try await URLSession.shared.data(for: request)
+            
+            guard let httpResponse = response as? HTTPURLResponse,
+                    httpResponse.statusCode == 200,
+                    var dataDictionary = try JSONSerialization.jsonObject(with: data, options: []) as? [[String: Any]] else {
+                return nil
```

Comment:
If during the pipeline a test fails, the engineer will look at the logs to see what happened. I think it would be useful to print rich logs. 

For example, a test might fail because the 1sec service is performing some throttling. Or because their service is down. I think printing the status code in case of failure might be interesting.

Also as we discussed before, the email can be empty, in which case is a Microsoft issue on delivering the email OTP. You can add that info to the print (if the body is an empty array, say that Microsoft didn't delivered the email OTP) so we will know exactly why that test failed.

Reply:
The goal is that a test should never fails for something related to email OTP code retriever. If an email OTP is not found the test is retried with a different email address. 
I added some basic log in the `retrieveEmailOTPCode` method. I didn't add logs in this method because it can be executed up to three times for each test execution, so I didn't log anything to don't create too much noise. 
We can talk more about it in a call if you want.